src/odemis/driver/piezomotor.py

```diff
@@ -0,0 +1,453 @@
+        HwComponent.__init__(self, name, role, *args, **kwargs)
+        Axis.__init__(self, canAbs=canAbs, choices=choices, unit=unit,
+                      range=range, speed=speed, canUpdate=canUpdate)
+
+
+class PMD401(Actuator):
+    """
+    This represents the PMD401 motor controller for the PiezoMotor LEGS motors.
+    The specification for the hardware interface can be found in the document
+     "PiezoMotor_PMD401_Technical_Manual.pdf".
+
+    The class provides only part of the functionality of the model.Actuator interface. It fully
+    implements the functions moveRel and stop. However, there is no reference function and
+    moveAbs is only possible if we are operating in closed loop, i.e. an encoder (stage position sensor)
+    is present.
+    """
+
+    def __init__(self, name, role, port, axes, ax_numbers, inverted=None, closed_loop=True, speed=1e-6, **kwargs):
+        """
+        :arg ax_numbers (dict: str --> int): axis name --> axis number
+        :arg closed loop (bool): True if stage position sensor is present, False otherwise
+        :arg speed (float): speed in m/s
+        """
+        # It is also possible to send a command to the device that lets the firmware deal with
+        # inverted axes directly, but since we already have code for this in odemis, we'll
+        # invert the axes in the software.
+        Actuator.__init__(self, name, role, axes=axes, inverted=inverted, **kwargs)
+        self._executor = CancellableThreadPoolExecutor(max_workers=1)  # one task at a time
+        self._ser_access = threading.RLock()
+
+        self._ax_numbers = ax_numbers
+        self.closed_loop = closed_loop
+        self._speed = speed
+
+        # Connect to hardware
+        self._serial = self._findDevice(port)
+
+        # Configuration
+        self.setWaveform(WAVEFORM_DELTA)
+
+    def stop(self, axes=None):
+        self._executor.cancel()
+        axes = axes or self.axes
+        for ax in axes:
+            self._sendCommand(b'X%dS' % self._ax_numbers[ax])
+
+    def moveRel(self, shift):
+        if not shift:
+            return model.InstantaneousFuture()
+        self._checkMoveRel(shift)
+        shift = self._applyInversion(shift)
+        f = self._executor.submit(self._doMoveRel, shift)
+        return f
+
+    def moveAbs(self, pos):
+        self._checkMoveAbs(pos)
+        pos = self._applyInversion(pos)
+        f = self._executor.submit(self._doMoveAbs, pos)
+        return f
+
+    def _doMoveAbs(self, pos):
+        if not self.closed_loop:
+            raise ValueError("Cannot perform absolute move in open loop mode.")
+
+        targets = {}
+        for axis, val in pos.items():
+            encoder_cnts = round(val / ENCODER_STEPSIZE)
+            speed_usteps = round(self._speed / WFM_STEPSIZE * USTEPS_PER_WFM)  # steps / second
+            self._sendCommand(b'X%dT%d,%d' % (self._ax_numbers[axis], encoder_cnts, speed_usteps))
+            targets[axis] = encoder_cnts
```

Comment:
Regularly, it'll need to check the status of the hardware (U0). At least, it should be done after each move (or while waiting for the move to finish).